### ISSUE TITLE
Variables / class attributes shouldn’t shadow types

### ISSUE THREAD

**Issue by flying-sheep** [astral-sh/ty#2699]
### Summary

In `numpy/random/generator.pyi`, we find

```py
class Generator:
    ...
    def bytes(self, length: int) -> bytes: ...
```

ty highlights the return type as function and ctrl-clicking it puts the cursor on the `def`, so it seems to think that the return type is identical with the function object, but it should be `builtins.bytes`

### Version

0.0.14

**Comment by sharkdp** [comment]
Thank you for reporting this.

Please see https://github.com/astral-sh/ty/issues/900 and https://github.com/astral-sh/ty/issues/2430 for related issues. https://github.com/astral-sh/ty/issues/900#issuecomment-3151616356 describes our reasoning and our general advise is to avoid shadowing names: https://github.com/astral-sh/ty/issues/2430#issuecomment-3731059444

**Comment by flying-sheep** [comment]
Ah I see they actually fixed it: https://github.com/numpy/numpy/pull/30754

I’m just a bit confused: functions aren’t types right? Otherwise I spent time arduously defining callable protocols for nothing.

**Comment by sharkdp** [comment]
> I’m just a bit confused: functions aren’t types right?

They are not, no. Functions are not valid in type expressions. That's why ty issues the `invalid-type-form` diagnostic here.

> Otherwise I spent time arduously defining callable protocols for nothing.

👍 

Only partially related... and just in case you're curious: the type *of* a function object in ty is a special singleton type (with just one inhabitant, the function object itself). It can't be spelled using normal Python typing syntax, but you can access it using `ty_extensions.TypeOf[f]`. If you annotate something with this type, only the function `f` itself would be acceptable. We also provide `ty_extensions.CallableTypeOf` which turns the type of `f` into a corresponding `Callable` supertype. See [this example](https://play.ty.dev/367c0069-25c1-4c19-ac0e-acaf29d24c64) for a demonstration.

But I wouldn't recommend using that in production. Spelling out the `Callable` type yourself or defining a callable protocol is the way to go.
